src/v/cloud_io/remote.h

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "cloud_io/auth_refresh_bg_op.h"
+#include "cloud_io/io_resources.h"
+#include "cloud_io/io_result.h"
+#include "cloud_io/transfer_details.h"
+#include "cloud_roles/refresh_credentials.h"
+#include "cloud_storage_clients/client.h"
+#include "cloud_storage_clients/client_pool.h"
+#include "model/metadata.h"
+#include "utils/lazy_abort_source.h"
+#include "utils/retry_chain_node.h"
+#include "utils/stream_provider.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/sharded.hh>
+#include <seastar/util/noncopyable_function.hh>
+
+#include <ranges>
+#include <utility>
+
+namespace cloud_io {
+
+struct upload_request {
+    transfer_details transfer_details;
+    std::string_view display_str;
+    iobuf payload;
+    bool accept_no_content_response{false};
+};
+
+struct download_request {
+    transfer_details transfer_details;
+    std::string_view display_str;
+    iobuf& payload;
+    bool expect_missing{false};
+};
+using remote_path = named_type<ss::sstring, struct remote_path_tag>;
+
+using list_result = result<
+  cloud_storage_clients::client::list_bucket_result,
+  cloud_storage_clients::error_outcome>;
+
+/// Functor that attempts to consume the input stream. If the connection
+/// is broken during the download the functor is responsible for he cleanup.
+/// The functor should be reenterable since it can be called many times.
+/// On success it should return content_length. On failure it should
+/// allow the exception from the input_stream to propagate.
+using try_consume_stream = ss::noncopyable_function<ss::future<uint64_t>(
+  uint64_t, ss::input_stream<char>)>;
+
+/// \brief Represents remote endpoint
+///
+/// The `remote` is responsible for remote data
+/// transfer and storage. It 'knows' how to upload and
+/// download data. Also, it's responsible for maintaining
+/// correct naming in S3. The remote takes into account
+/// things like reconnects, backpressure and backoff.
+class remote : public ss::peering_sharded_service<remote> {
```

Comment:
should we use this opportunity to give this thing a better name?

Reply:
I'm open to it, for sure. I was thinking `remote_io` or something.

Also may opt to do this in a follow-up just so we can start using this code